riscv/csrs.cc

```diff
@@ -454,13 +454,19 @@ sstatus_csr_t::sstatus_csr_t(processor_t* const proc, sstatus_proxy_csr_t_p orig
 }
 
 void sstatus_csr_t::dirty(const reg_t dirties) {
+  // As an optimization, return early if already dirty.
+  if ((orig_sstatus->read() & dirties) == dirties) {
+    if (likely(!state->v || (virt_sstatus->read() & dirties) == dirties))
+      return;
+  }
+
   // Catch problems like #823 where P-extension instructions were not
   // checking for mstatus.VS!=Off:
   if (!enabled(dirties)) abort();
 
-  orig_csr->write(orig_csr->read() | dirties);
+  orig_sstatus->write(orig_sstatus->read() | dirties);
   if (state->v) {
-    virt_csr->write(virt_csr->read() | dirties);
+    virt_sstatus->write(virt_sstatus->read() | dirties);
```

Comment:
Correct re: vtable.  I actually could've left these alone since the first part of the commit makes this code rarely executed.  But since the first part of the commit needs to access `orig_sstatus` and `virt_sstatus` for perf reasons, I thought it looked confusing to use `orig_csr` and `virt_csr` later on in the same method.

I didn't realize I was removing side effects by removing the writes.  Thanks for pointing that out; I'll split out that commit.